src/Ouroboros/Network/ChainFragment.hs

```diff
@@ -0,0 +1,540 @@
+--
+-- * The 'blockPrevHash' of the @bSucc@ must match that of @b@.
+-- * The 'blockSlot' of @bSucc@ must be strictly larger than that of @b@.
+-- * The 'blockNo' of @bSucc@ must be 1 greater than that of @b@.
+--
+-- This function does not check whether any of the two blocks satisfy
+-- 'blockInvariant'.
+isValidSuccessorOf :: HasHeader block
+                   => block  -- ^ @bSucc@
+                   -> block  -- ^ @b@
+                   -> Bool
+isValidSuccessorOf bSucc b =
+    pointHash p == blockPrevHash bSucc
+ && pointSlot p <  blockSlot bSucc
+ && blockNo   b == pred (blockNo bSucc)
+  where
+    p = blockPoint b
+
+validExtension ::  HasHeader block => ChainFragment block -> block -> Bool
+validExtension c bSucc = blockInvariant bSucc && case head c of
+  Nothing -> True
+  Just b  -> bSucc `isValidSuccessorOf` b
+
+head :: HasHeader block => ChainFragment block -> Maybe block
+head (_ :> b) = Just b
+head Empty    = Nothing
+
+headPoint :: HasHeader block => ChainFragment block -> Maybe (Point block)
+headPoint (_ :> b) = Just (blockPoint b)
+headPoint Empty    = Nothing
+
+headSlot :: HasHeader block => ChainFragment block -> Maybe Slot
+headSlot c = pointSlot <$> headPoint c
+
+headHash :: HasHeader block => ChainFragment block -> Maybe (Hash block)
+headHash c = pointHash <$> headPoint c
+
+headBlockNo :: HasHeader block => ChainFragment block -> Maybe BlockNo
+headBlockNo (_ :> b) = Just (blockNo b)
+headBlockNo Empty    = Nothing
+
+
+-- | Make a list of blocks from a 'ChainFragment', in newest-to-oldest order.
+toNewestFirst :: HasHeader block => ChainFragment block -> [block]
+toNewestFirst = foldChainFragment (flip (:)) []
+
+-- | Make a list of blocks from a 'ChainFragment', in oldest-to-newest order.
+toOldestFirst :: HasHeader block => ChainFragment block -> [block]
+toOldestFirst (ChainFragment ft) = Foldable.toList ft
+
+-- | Make a 'ChainFragment' from a list of blocks in newest-to-oldest order.
+fromNewestFirst :: HasHeader block => [block] -> ChainFragment block
+fromNewestFirst bs = foldr (flip (:>)) Empty bs
+
+-- | Make a 'ChainFragment' from a list of blocks in oldest-to-newest order.
+fromOldestFirst :: HasHeader block => [block] -> ChainFragment block
+fromOldestFirst bs = assert (valid c) c
+  where
+    c = ChainFragment $ FT.fromList bs
+
+-- | Drop the last @n@ blocks from the 'ChainFragment'.
+drop :: HasHeader block
+     => Int  -- ^ @n@
+     -> ChainFragment block -> ChainFragment block
+drop n cf@(ChainFragment c) =
+    ChainFragment $ FT.takeUntil (\v -> bmSize v > remainingLength) c
+  where
+    remainingLength = length cf - n
+
+-- | Drop the oldest/first @n@ blocks from the 'ChainFragment'.
+dropOldest :: HasHeader block
+           => Int  -- ^ @n@
+           -> ChainFragment block -> ChainFragment block
+dropOldest n (ChainFragment c) =
+    ChainFragment $ FT.dropUntil (\v -> bmSize v > n) c
+
+
+length :: ChainFragment block -> Int
+length (ChainFragment c) = Foldable.length c
```

Comment:
Is there a reason _not_ to declare `ChainFragment` itself to be foldable? Then all these wrappers around `Foldable.stuff` would not be necessary right?

Reply:
My reasoning is that the exposed wrappers are explicit about the ordering (right-to-left or the other way around, or oldest first, etc). Just exposing the `Foldable` instance would shift the burden of thinking about this to the users.

But I don't mind adding a `Foldable (ChainFragment block)` instance if you like.

Reply:
Aha, gotcha. Yup, sounds reasonable.